### PR TITLE
User Cart

### PR DESCRIPTION
**Pull request by brianwchou**
Create an endpoint POST /users/:id/cart
        Enter data when:
            - user has a cart (edit cart)
            - user doesnt have a cart (create cart)

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1,11 +1,45 @@
...
+// user wants to buy a book
+app.post('/users/:userId/cart/:cartId', (req, res) => {
+  const { userId, cartId } = req.params;
+  const booklist = req.body;
+
+  // check if the books are in the database
+
+  let bookIndexes = [];
+  for (let bookTitle of booklist) {
+    let titleIsFound = bookDB.some((bookEntry) => {
```
Comment (kveryha) [inline index.js]: `titleIsFound` - great naming!

File: index.js [hunk 2]
```diff
@@ -1,11 +1,45 @@
...
+      return bookEntry.title === bookTitle;
```
Comment (kveryha) [inline index.js]: in case u have pretty simple arrow function , then shorthand can be used `bookDB.some(({ title }) => title === bookTitle)`

File: index.js [hunk 3]
```diff
@@ -1,11 +1,45 @@
 const express = require('express');
 const bodyParser = require('body-parser');
 const { bookRoutes, errorRoutes, customerRoutes } = require('./routers');
+const { bookDB, cartDB } = require('./db');
 const app = express();
 const port = 8080;
 
 app.use(bodyParser.json());
 
+// user wants to buy a book
```
Comment (kveryha) [inline index.js]: Here we are! This is a great example why do we need splitting parts of the code.
You have a comment here which explains what is done here. And this is  kind of a sign to move things to a separate piece of code, in this case it should be a function.
At the moment your approach requires a reader to spend more time while reviewing.
Could you please move handler part into separate functions and use them here? If you do it correct, the comment won't be necessary.

File: index.js [hunk 4]
```diff
@@ -1,11 +1,45 @@
...
+      res.status(400).send(`Could not find book title: ${bookTitle}`);
+      return;
+    }
+
+    // find book index
+    for (let i = 0; i < bookDB.length; i++) {
+      if (bookDB[i].title === bookTitle) {
+        bookIndexes.push(i);
+      }
+    }
```
Comment (kveryha) [inline index.js:33]: At this point I don't see response status...

File: index.js (hunk 3 above)
Comment (brianwchou) [reply index.js]: It's funny you point this out because as soon as I finished writing my function out I started to see the need for creating the service layer abstraction. I'll be sure to have this reorganized in another commit.

File: index.js (hunk 2 above)
Comment (brianwchou) [reply index.js]: Nice catch!

File: index.js (hunk 3 above)
Comment (kveryha) [reply index.js]: Yeah, thanks for pointing me that it's still a draft!

File: repositories/books.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
+const { bookDB } = require('../db');
+
+function addBookWithAuthorTitleAndPages(author, title, pages) {
```
Comment (kveryha) [inline repositories/books.js]: _too_ descriptive naming here))
`addBookWithAuthorTitleAndPages(author, title, pages)` should be 
`addBook(author, title, pages) `or `addBook({author, title, pages})`

File: repositories/books.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+function deleteBookWithId(id) {
+  if (id === 0) {
+    bookDB = bookDB.slice(1, bookDB.length);
+  } else {
+    let firsthalf = bookDB.slice(0, id);
+    let secondhalf = bookDB.slice(id + 1, bookDB.length);
+
+    bookDB = [...firsthalf, ...secondhalf];
+  }
```
Comment (kveryha) [inline repositories/books.js]: Let's just use `splice` method here
Anyway, until it's a in-memory DB you need to modify the array
In the future when we move to real DB engine these things will go

File: repositories/users.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+  });
+}
+
+function getUserRecord(firstName, lastName) {
+  return userDB.find((entry) => {
+    if (entry.firstName === firstName || entry.lastName === lastName) {
+      return entry;
+    }
+  });
+}
```
Comment (kveryha) [inline repositories/users.js]: It's ok here, but FYI usually it's common to rely on id or at least email + password data

File: routers/books.js [hunk 8]
```diff
@@ -2,18 +2,18 @@ const { Router } = require('express');
...
 const {
   deleteBooks,
-  addBookEntry,
+  addBook,
   getAllBooks,
-  getBookEntry,
-  updateBookEntry,
+  getBook,
+  updateBook,
 } = require('../controllers/books');
```
Comment (kveryha) [inline routers/books.js]: What I see that you mixed concepts a little bit. 
TL;DR controller = router
to stay with Express notation, let's call it router

Comment (kveryha) [reply routers/books.js]: (this comment is not finished yet)

File: services/user.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
+const { userRepository } = require('../repositories');
+
+function findUserByName(firstName, lastName) {
+  return userRepository.getUserRecord(firstName, lastName);
+}
+
+function createNewUser(username, firstName, lastName, email, password) {
```
Comment (kveryha) [inline services/user.js]: too many params here. when it comes to 3 or more, it's better to pass objects

**Review by kveryha — COMMENTED** [review]
reviewing WIP

File: routers/books.js (hunk 8 above)
Comment (brianwchou) [reply routers/books.js]: I see what you mean now, I always assumed that they were separate. **Routing** is part of the controller layer

File: controllers/books.js [hunk 10]
```diff
@@ -1,46 +1,48 @@
-const { bookDB } = require('../db');
+const { bookRepository } = require('../repositories');
```
Comment (kveryha) [inline controllers/books.js]: rename file to `book.js`

File: controllers/user.js [hunk 11]
```diff
@@ -0,0 +1,45 @@
+const { usercartService, userService } = require('../services');
```
Comment (kveryha) [inline controllers/user.js]: rename `usercartService` to `userCartService`

File: controllers/books.js [hunk 12]
```diff
@@ -1,46 +1,48 @@
...
+
+  bookRepository.addBookWithAuthorTitleAndPages(author, title, pages);
 
-const addBookEntry = (req, res) => {
-  const newBookEntry = req.body;
-  bookDB.push(newBookEntry);
   res.send('record creation sucessful').status(201);
 };
 
 const getAllBooks = (req, res) => {
```
Comment (kveryha) [inline controllers/books.js]: it's more common to use function declaration
[explanation](https://javascript.info/function-expressions#function-expression-vs-function-declaration)

File: routers/user.js [hunk 13]
```diff
@@ -2,13 +2,13 @@ const { Router } = require('express');
...
   getUserByName,
-  createUser,
+  addUser,
   putBooksInCart,
 } = require('../controllers/user');
 const { userValidation } = require('../middlewares');
 
 userRouter.get('/find', getUserByName);
-userRouter.post('/create', createUser);
+userRouter.post('/add', addUser);
```
Comment (kveryha) [inline routers/user.js]: Due to explicit semantics behind the HTTP method names you don't need to have `userRouter.get('/find', getUserByName)`, here `userRouter.get('/', getUserByName)` will go

Comment (kveryha) [reply routers/user.js]: the same for `POST`. By using POST you're saying that you want to create a resource

Comment (brianwchou) [reply routers/user.js]: Thanks Kate, I can see having those verbs can be redundant

Comment (kveryha) [reply routers/user.js]: 👍

File: routers/user.js [hunk 14]
```diff
@@ -0,0 +1,48 @@
+const { Router } = require('express');
+const userRouter = Router();
```
Comment (kveryha) [inline routers/user.js]: could you please move router declaration a few lines down separately from `require` statements?

File: index.js [hunk 15]
```diff
@@ -1,13 +1,14 @@
 const express = require('express');
 const bodyParser = require('body-parser');
-const { bookRoutes, errorRoutes, customerRoutes } = require('./routers');
+
+const { bookRoutes, errorRoutes, userRoutes } = require('./routers');
```
Comment (kveryha) [inline index.js]: it would be better to give names to imports the same (or, at least pretty similar) as in original modules. So u have there `bookRouter`, `userRouter` etc
